### ISSUE TITLE
Optimum number of mpi processes and openmp threads for hyperthreaded Intel Xeon Processor E5-2650 v4 nodes

### ISSUE THREAD

**Issue by GeorgeHicks1** [SmileiPIC/Smilei#71]
Hi,

I am running smilei on one Intel Xeon Processor E5-2650 v4 node. I am told by the ICT support that "these nodes have 12 cores and 2 threads per core (24). There are 2 processor sockets per node."
Therefore I can have 48 mpi processes or 48 omp therads or something in between.

What is the optimum? I believe it should be 2 mpi processses to do the inter-socket communication and 24 omp threads to do the intra-socket communication. Is this correct?

On a related note I also use multiple (say 8) Intel Xeon Processor X5650 nodes. There have 6 cores and 2 threads per core. Therefore should I use 1 mpi process per node and 12 omp threads?

Thanks for your help,
George Hicks

**Comment by jderouillat** [comment]
Hi,
*I believe it should be 2 mpi processes to do the inter-socket communication* : right, this is the main point. To keep memory locality, threads may not go further that the socket which owns the mpi process.
*24 omp threads* : I recommend to not use hyper-threads (so 12) or at least to compare on a smaller case that your targeted one with and without hyper-threads. It's the same for the X5650.

Another recommendation is that if the simulation do not use a large number of cores favor MPI only configurations, on the contrary if you use a large number of cores, your described configuration is the good one. 

Imbalance is managed through the LoadBalancing module of the code, which works as fine in MPI only as in MPI x OpenMP.  But the main improvement in the MPI x OpenMP configuration consists in the behavior of diagnostics which scale more efficiently.
At a small scale, the interleaved parallelism levels have an overhead which is not counterbalanced by the diagnostic point.

Julien

**Comment by mccoys** [comment]
To add to @jderouillat recommendations, I must emphasize that you should *test* these configurations before accepting the conclusions. You get surprises sometimes.

**Comment by GeorgeHicks1** [comment]
Hi @jderouillat and @mccoys,

Thank you for your advice. I have carried out some testing and it appears that favouring more mpi processes and fewer open mp threads, speed-ups being around 7%. There is also a 3% speed-up from using `export OMP_SCHEDULE = DYNAMIC`.

I will share the full results when the last simulations have finished later today or tomorrow.

Thanks,
George

**Comment by GeorgeHicks1** [comment]
Here is a summary of my testing.

MPI processes | OpenMP threads | Running time/ hours
------------ | -------------  | -------------
2 | 24 | 3.11
4 | 12 | 3.02
12 | 4 | 2.97
24 | 2 | 2.90
48 | 1 | 3.01

The optimum MPI/OMP distribution is 24/2. This, as @mccoys suggested, did surprise me! The hardware I am using is detailed in my first post. I am simulating a thin-foil in a 12800x1344 box, with linearized MPI patches and 90e6 particles.

All these simulations were carried out with `export OMP_SCHEDULE = DYNAMIC`. I did a comparison simulation with out this for the 4 MPI process/ 12 OMP thread case which took 3.12 hours, a slow down of ~3% as I stated in my previous post.
